pkgs/top-level/default.nix

```diff
@@ -57,24 +56,36 @@ let
   platform = if platform_ != null then platform_
     else config.platform or platformAuto;
 
-  topLevelArguments = {
-    inherit system bootStdenv noSysDirs config crossSystem platform lib;
-  };
+  # A few packages make a new package set to draw their dependencies from.
+  # (Currently to get a cross tool chain, or forced-i686 package.) Rather than
+  # give `all-packages.nix` all the arguments to this function, even ones that
+  # don't concern it, we give it this function to "re-call" nixpkgs, inheriting
+  # whatever arguments it doesn't explicitly provide. This way,
+  # `all-packages.nix` doesn't know more than it needs too.
+  #
+  # It's OK that `args` doesn't include the defaults: they'll be
+  # deterministically inferred the same way.
+  mkPackages = newArgs: import ./. (args // newArgs);
 
   stdenvAdapters = self: super:
     let res = import ../stdenv/adapters.nix self; in res // {
       stdenvAdapters = res;
     };
 
   trivialBuilders = self: super:
-    (import ../build-support/trivial-builders.nix {
+    import ../build-support/trivial-builders.nix {
       inherit lib; inherit (self) stdenv stdenvNoCC; inherit (self.xorg) lndir;
-    });
+    };
 
-  stdenvDefault = self: super: (import ./stdenv.nix topLevelArguments) {} pkgs;
```

Comment:
@nbp So yeah, my hope is to get this one merged as is, as multiple people have reviewed it positively, and your only complaint is procedural. Then the next one sets up cross-compiling correctly, but its probably still broken in practice (it's subtly broken already). Finally using @DavidEGrayson's and @vcunat's good work atop my foundation, we can actually fix cross-compiling.